x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/LookupJoin.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.plan.logical.join;
+
+import org.elasticsearch.xpack.esql.core.expression.Attribute;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.plan.logical.Project;
+import org.elasticsearch.xpack.esql.plan.logical.SurrogateLogicalPlan;
+import org.elasticsearch.xpack.esql.plan.logical.join.JoinTypes.UsingJoinType;
+
+import java.util.List;
+import java.util.Objects;
+
+import static java.util.Collections.emptyList;
+import static org.elasticsearch.xpack.esql.plan.logical.join.JoinTypes.LEFT;
+
+/**
+ * Lookup join - specialized LEFT (OUTER) JOIN between the main left side and a lookup index (index_mode = lookup) on the right.
+ * In the future, as the join capabilities of the engine will evolve, a regular LEFT JOIN will be used instead, letting the planner decide
+ * the strategy at runtime.
+ */
+public class LookupJoin extends Join implements SurrogateLogicalPlan {
+
+    private final List<Attribute> output;
+
+    public LookupJoin(Source source, LogicalPlan left, LogicalPlan right, List<Attribute> joinFields) {
+        this(source, left, right, new JoinConfig(new UsingJoinType(LEFT, joinFields), emptyList(), emptyList(), emptyList()), emptyList());
+    }
+
+    public LookupJoin(Source source, LogicalPlan left, LogicalPlan right, JoinConfig joinConfig, List<Attribute> output) {
+        super(source, left, right, joinConfig);
+        this.output = output;
+    }
+
+    /**
+     * Translate the expression into a regular join with a Projection on top, to deal with serialization &amp; co.
+     */
+    @Override
+    public LogicalPlan surrogate() {
+        JoinConfig cfg = config();
+        JoinConfig newConfig = new JoinConfig(LEFT, cfg.matchFields(), cfg.leftFields(), cfg.rightFields());
+        Join normalized = new Join(source(), left(), right(), newConfig);
+        // TODO: decide whether to introduce USING or just basic ON semantics - keep the ordering out for now
+        return new Project(source(), normalized, output);
+    }
+
+    public List<Attribute> output() {
+        return output;
+    }
```

Comment:
I've added a line in #116208.
The main reason why right now it's in the Analyzer is because in case of failure, the LookupJoin becomes an unresolved Join since it doesn't survive (it gets rewritten to a regular join with a special output due to the particularities of using).
Separately the logic around LEFT/RIGHT, etc.. types of join and dealing with each branch are harder to encapsulate because the class has to check itself whether it's children are resolved, what's left and right side.
This type of pattern matching and traversal is done outside the tree classes.

See the TODO I added in JoinConfig and my previous comments.
JoinConfig models an equijoin however moving forward a condition expression since it encapsulates better the generality in the future (ON a > b) and simplifies tree traversal (JoinConfig doesn't extend Node and thus cannot be visited by our rules).